include/submission_functions.php

```diff
@@ -343,6 +326,48 @@ function ProcessSubmissions($projectid)
 }
 
 
+function GetNextSubmission($projectid)
+{
+    $now_utc = gmdate(FMT_DATETIMESTD);
+
+    // Lock the table to avoid a race condition between multiple processes
+    // fighting over the same submission.
+    $tables = array();
+    $tables[] = 'submission';
+    if (!pdo_lock_tables($tables)) {
+        add_log("could not lock submission table", "GetNextSubmission",
+                LOG_ERR, $projectid);
+        return false;
+    }
+
+    // Get the next submission to process.
+    $query_array = pdo_single_row_query(
+            "SELECT id, filename, filesize, filemd5sum, attempts
+            FROM submission
+            WHERE projectid='$projectid' AND status=0
+            ORDER BY id LIMIT 1");
+    add_last_sql_error("GetNextSubmission-1");
+
+    $submission_id = $query_array['id'];
+    $new_attempts = $query_array['attempts'] + 1;
+
+    // Mark it as status=1 (processing) and record started time.
+    pdo_query("UPDATE submission SET status=1, started='$now_utc', ".
+            "lastupdated='$now_utc', attempts=$new_attempts ".
+            "WHERE id='".$submission_id."'");
+    add_last_sql_error("GetNextSubmission-2");
+
+    // Unlock the submission table.
+    if (!pdo_unlock_tables($tables)) {
+        add_log("could not unlock submission table", "GetNextSubmission",
+                LOG_ERR, $projectid);
+        return false;
+    }
```

Comment:
I am sure there is something better, but if each worker had an ID, you could query that the ID had indeed been updated to that ID. The update would also need a where worker ID not set clause to avoid the ID being overwritten. The database should naturally handle two simultaneous updates otherwise.